File: app/packages/core/src/ur.ts
/**
 * Uniform Resources (BC-UR): the wire format for the QR air gap.
 *
 * The mirror of `src/ur.c`, in the same relationship `cbor.ts` has to
 * `src/cbor.c` — two implementations of one wire format, held together by
 * vectors the firmware emits. `sim/test_ur.c --emit-vectors` writes
 * `test/ur-vectors.json`, `scripts/check.sh` regenerates it before this
 * package's tests run, and `test/ur.test.ts` replays it. A corpus recorded
 * from an older `ur.c` would let this file agree with an encoder nobody runs,
 * which is the failure docs/MIRROR-GAP.md describes.
 *
 * BC-UR is the encoding: CBOR payload, CRC-32, Bytewords text, and a fountain
 * code for splitting across animated frames. EIP-4527 is a profile on top that
 * names the CBOR structures (`crypto-hdkey`, `eth-sign-request`,
 * `eth-signature`) and says nothing about the encoding. They are kept apart
 * here for the same reason they are in the firmware: 4527 is marked Stagnant
 * while BC-UR is what wallets actually interoperate over.
 *
 * The word list is data from Blockchain Commons' bc-ur, BSD-2-Clause Plus
 * Patent. The implementation is ours.
 */

/**
 * The 256 Bytewords, concatenated, four characters each.
 *
 * Copied verbatim rather than retyped: the first and last letters of each word
 * ARE the encoding, so one transposed letter would produce output other
 * wallets decode to different bytes, and the checksum would not catch it
 * because it would be computed over the same wrong understanding.
 */
const BYTEWORDS =
  "ableacidalsoapexaquaarchatomauntawayaxisbackbaldbarnbeltbetabias" +
  "bluebodybragbrewbulbbuzzcalmcashcatschefcityclawcodecolacookcost" +
  "cruxcurlcuspcyandarkdatadaysdelidicedietdoordowndrawdropdrumdull" +
  "dutyeacheasyechoedgeepicevenexamexiteyesfactfairfernfigsfilmfish" +
  "fizzflapflewfluxfoxyfreefrogfuelfundgalagamegeargemsgiftgirlglow" +
  "goodgraygrimgurugushgyrohalfhanghardhawkheathelphighhillholyhope" +
  "hornhutsicedideaidleinchinkyintoirisironitemjadejazzjoinjoltjowl" +
  "judojugsjumpjunkjurykeepkenokeptkeyskickkilnkingkitekiwiknoblamb" +
  "lavalazyleaflegsliarlimplionlistlogoloudloveluaulucklungmainmany" +
  "mathmazememomenumeowmildmintmissmonknailnavyneednewsnextnoonnote" +
  "numbobeyoboeomitonyxopenovalowlspaidpartpeckplaypluspoempoolpose" +
  "puffpumapurrquadquizraceramprealredorichroadrockroofrubyruinruns" +
  "rustsafesagascarsetssilkskewslotsoapsolosongstubsurfswantacotask" +
  "taxitenttiedtimetinytoiltombtoystriptunatwinuglyundouniturgeuser" +
  "vastveryvetovialvibeviewvisavoidvowswallwandwarmwaspwavewaxywebs" +
  "whatwhenwhizwolfworkyankyawnyellyogayurtzapszerozestzinczonezoom";

/** first+last letter pair -> byte value */
const MINIMAL_TO_BYTE = new Map<string, number>();
for (let i = 0; i < 256; i++) {
  MINIMAL_TO_BYTE.set(BYTEWORDS[i * 4]! + BYTEWORDS[i * 4 + 3]!, i);
}

const CRC32_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let i = 0; i < 256; i++) {
    let c = i;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    t[i] = c >>> 0;
  }
  return t;
})();

/** Standard CRC-32 (IEEE 802.3: reflected, init and final xor 0xFFFFFFFF). */
export function crc32(data: Uint8Array): number {
  let crc = 0xffffffff;
  for (const b of data) {
    crc = CRC32_TABLE[(crc ^ b) & 0xff]! ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

/** Encode bytes as Bytewords-minimal, appending the CRC-32. */
export function bytewordsEncode(data: Uint8Array): string {
  const crc = crc32(data);
  const withCrc = new Uint8Array(data.length + 4);
  withCrc.set(data, 0);
  withCrc[data.length] = (crc >>> 24) & 0xff;
  withCrc[data.length + 1] = (crc >>> 16) & 0xff;
  withCrc[data.length + 2] = (crc >>> 8) & 0xff;
  withCrc[data.length + 3] = crc & 0xff;

  let out = "";
  for (const b of withCrc) {
    out += BYTEWORDS[b * 4]! + BYTEWORDS[b * 4 + 3]!;
  }
  return out;
}

/**
 * Decode Bytewords-minimal and verify the CRC-32.
 *
 * Throws rather than returning null on a bad frame. A misread QR is the
 * expected case, not an exceptional one, so callers scanning an animation are
 * meant to catch and keep scanning; what must never happen is a corrupted
 * frame decoding to plausible bytes.
 */
export function bytewordsDecode(text: string): Uint8Array {
  if (text.length % 2 !== 0) {
    throw new Error("bytewords: odd length");
  }
  if (text.length < 8) {
    throw new Error("bytewords: too short to carry a checksum");
  }

  const total = text.length / 2;
  const out = new Uint8Array(total);
  for (let i = 0; i < total; i++) {
    const pair = text.slice(i * 2, i * 2 + 2).toLowerCase();
    const v = MINIMAL_TO_BYTE.get(pair);
    if (v === undefined) {
      throw new Error(`bytewords: no word for "${pair}"`);
    }
    out[i] = v;
  }

  const payload = out.subarray(0, total - 4);
  const tail = out.subarray(total - 4);
  const want =
    ((tail[0]! << 24) | (tail[1]! << 16) | (tail[2]! << 8) | tail[3]!) >>> 0;
  if (crc32(payload) !== want) {
    throw new Error("bytewords: checksum mismatch");
  }
  return payload;
}

/** Longest UR type we accept, e.g. "eth-sign-request". */
export const UR_TYPE_MAX = 32;

function validType(type: string): boolean {
  if (type.length === 0 || type.length > UR_TYPE_MAX) return false;
  if (type.startsWith("-") || type.endsWith("-")) return false;
  return /^[a-z0-9-]+$/.test(type.toLowerCase());
}

/** Encode a single-part UR: `ur:<type>/<bytewords>`. */
export function urEncode(type: string, payload: Uint8Array): string {
  if (!validType(type)) {
    throw new Error(`ur: invalid type "${type}"`);
  }
  return `ur:${type.toLowerCase()}/${bytewordsEncode(payload)}`;
}

export interface UrDecoded {
  type: string;
  payload: Uint8Array;
}

/**
 * Decode a single-part UR.
 *
 * Accepts the scheme in any case: QR readers and some wallets uppercase the
 * whole string to reach the QR alphanumeric mode, which is smaller on screen.
 * Multi-part URs are rejected here rather than mis-parsed; see `isMultipart`.
 */
export function urDecode(ur: string): UrDecoded {
  const parts = ur.split("/");
  if (parts.length !== 2) {
    throw new Error("ur: expected exactly one '/'");
  }
  const head = parts[0]!;
  if (!/^ur:/i.test(head)) {
    throw new Error("ur: missing scheme");
  }
  const type = head.slice(3).toLowerCase();
  if (!validType(type)) {
    throw new Error(`ur: invalid type "${type}"`);
  }
  return { type, payload: bytewordsDecode(parts[1]!) };
}

/** True if this UR carries a sequence component and needs the fountain decoder. */
export function isMultipart(ur: string): boolean {
  return (ur.match(/\//g) ?? []).length > 1;
}

File: src/ur.c
/* Uniform Resources (BC-UR) encoding. See ur.h for the shape of this. */

#include "ur.h"

#include <string.h>

/*
 * The 256 Bytewords, concatenated, four characters each.
 *
 * Data from Blockchain Commons' bc-ur (BSD-2-Clause Plus Patent). Copied
 * verbatim rather than retyped: the first and last letters of each word are the
 * encoding, so a single transposed letter here would produce output that other
 * wallets decode to different bytes, and the CRC would not catch it because we
 * would have computed the CRC over the same wrong understanding.
 */
static const char BYTEWORDS[] =
    "ableacidalsoapexaquaarchatomauntawayaxisbackbaldbarnbeltbetabias"
    "bluebodybragbrewbulbbuzzcalmcashcatschefcityclawcodecolacookcost"
    "cruxcurlcuspcyandarkdatadaysdelidicedietdoordowndrawdropdrumdull"
    "dutyeacheasyechoedgeepicevenexamexiteyesfactfairfernfigsfilmfish"
    "fizzflapflewfluxfoxyfreefrogfuelfundgalagamegeargemsgiftgirlglow"
    "goodgraygrimgurugushgyrohalfhanghardhawkheathelphighhillholyhope"
    "hornhutsicedideaidleinchinkyintoirisironitemjadejazzjoinjoltjowl"
    "judojugsjumpjunkjurykeepkenokeptkeyskickkilnkingkitekiwiknoblamb"
    "lavalazyleaflegsliarlimplionlistlogoloudloveluaulucklungmainmany"
    "mathmazememomenumeowmildmintmissmonknailnavyneednewsnextnoonnote"
    "numbobeyoboeomitonyxopenovalowlspaidpartpeckplaypluspoempoolpose"
    "puffpumapurrquadquizraceramprealredorichroadrockroofrubyruinruns"
    "rustsafesagascarsetssilkskewslotsoapsolosongstubsurfswantacotask"
    "taxitenttiedtimetinytoiltombtoystriptunatwinuglyundouniturgeuser"
    "vastveryvetovialvibeviewvisavoidvowswallwandwarmwaspwavewaxywebs"
    "whatwhenwhizwolfworkyankyawnyellyogayurtzapszerozestzinczonezoom";

/* Reflected CRC-32, computed a nibble at a time: the 16-entry table costs 64
   bytes of flash against 1 KB for the byte-wise one, and nothing here is in a
   hot loop. */
static const uint32_t CRC32_NIBBLE[16] = {
    0x00000000, 0x1DB71064, 0x3B6E20C8, 0x26D930AC,
    0x76DC4190, 0x6B6B51F4, 0x4DB26158, 0x5005713C,
    0xEDB88320, 0xF00F9344, 0xD6D6A3E8, 0xCB61B38C,
    0x9B64C2B0, 0x86D3D2D4, 0xA00AE278, 0xBDBDF21C,
};

uint32_t ur_crc32(const uint8_t *data, size_t len)
{
    uint32_t crc = 0xFFFFFFFFu;
    for (size_t i = 0; i < len; i++) {
        crc ^= data[i];
        crc = (crc >> 4) ^ CRC32_NIBBLE[crc & 0x0F];
        crc = (crc >> 4) ^ CRC32_NIBBLE[crc & 0x0F];
    }
    return crc ^ 0xFFFFFFFFu;
}

static char lower(char c)
{
    return (c >= 'A' && c <= 'Z') ? (char)(c - 'A' + 'a') : c;
}

/* Byte value for a first/last letter pair, or -1 if no word has that pair.
   Linear over 256 words: a decode of a full frame is a few tens of thousands of
   character compares, which is nothing beside the QR decode that produced it,
   and it costs no table and no lazy initialisation. */
static int byteword_lookup(char first, char last)
{
    first = lower(first);
    last = lower(last);
    for (int i = 0; i < 256; i++) {
        if (BYTEWORDS[i * 4] == first && BYTEWORDS[i * 4 + 3] == last) {
            return i;
        }
    }
    return -1;
}

static size_t encode_minimal(const uint8_t *data, size_t len, char *out)
{
    for (size_t i = 0; i < len; i++) {
        out[i * 2]     = BYTEWORDS[data[i] * 4];
        out[i * 2 + 1] = BYTEWORDS[data[i] * 4 + 3];
    }
    return len * 2;
}

size_t ur_bytewords_encode(const uint8_t *data, size_t len,
                           char *out, size_t out_size)
{
    const size_t need = UR_BYTEWORDS_LEN(len);
    if (out == NULL || out_size < need + 1) {
        return 0;
    }
    if (len > 0 && data == NULL) {
        return 0;
    }

    size_t n = encode_minimal(data, len, out);

    /* The CRC is appended big-endian and encoded the same way as the payload,
       so a decoder never needs to know where the payload stopped. */
    const uint32_t crc = ur_crc32(data, len);
    const uint8_t tail[4] = {
        (uint8_t)(crc >> 24), (uint8_t)(crc >> 16),
        (uint8_t)(crc >> 8),  (uint8_t)crc,
    };
    n += encode_minimal(tail, sizeof tail, out + n);

    out[n] = '\0';
    return n;
}

bool ur_bytewords_decode(const char *text, size_t text_len,
                         uint8_t *out, size_t out_size, size_t *out_len)
{
    if (text == NULL || out_len == NULL) {
        return false;
    }
    /* Two characters per byte, and at least the four CRC bytes must be there. */
    if ((text_len % 2) != 0 || text_len < 8) {
        return false;
    }

    const size_t total = text_len / 2;
    const size_t payload = total - 4;
    if (payload > out_size) {
        return false;
    }

    uint8_t crc_bytes[4];
    for (size_t i = 0; i < total; i++) {
        const int v = byteword_lookup(text[i * 2], text[i * 2 + 1]);
        if (v < 0) {
            return false;
        }
        if (i < payload) {
            out[i] = (uint8_t)v;
        } else {
            crc_bytes[i - payload] = (uint8_t)v;
        }
    }

    const uint32_t want = ((uint32_t)crc_bytes[0] << 24) |
                          ((uint32_t)crc_bytes[1] << 16) |
                          ((uint32_t)crc_bytes[2] << 8)  |
                          (uint32_t)crc_bytes[3];
    if (ur_crc32(out, payload) != want) {
        return false;
    }

    *out_len = payload;
    return true;
}

/* UR types are lowercase letters, digits and hyphens, and may not start or end
   with a hyphen. Checked rather than assumed, because the type is echoed into
   whatever the caller does next. */
static bool valid_type(const char *type, size_t len)
{
    if (len == 0 || len > UR_TYPE_MAX) {
        return false;
    }
    if (type[0] == '-' || type[len - 1] == '-') {
        return false;
    }
    for (size_t i = 0; i < len; i++) {
        const char c = lower(type[i]);
        const bool ok = (c >= 'a' && c <= 'z') ||
                        (c >= '0' && c <= '9') || c == '-';
        if (!ok) {
            return false;
        }
    }
    return true;
}

size_t ur_encode(const char *type, const uint8_t *payload, size_t payload_len,
                 char *out, size_t out_size)
{
    if (type == NULL || out == NULL) {
        return 0;
    }
    const size_t type_len = strlen(type);
    if (!valid_type(type, type_len)) {
        return 0;
    }

    /* "ur:" + type + "/" + body + NUL */
    const size_t need = 3 + type_len + 1 + UR_BYTEWORDS_LEN(payload_len);
    if (out_size < need + 1) {
        return 0;
    }

    memcpy(out, "ur:", 3);
    for (size_t i = 0; i < type_len; i++) {
        out[3 + i] = lower(type[i]);
    }
    out[3 + type_len] = '/';

    const size_t body = ur_bytewords_encode(payload, payload_len,
                                            out + 4 + type_len,
                                            out_size - 4 - type_len);
    if (body == 0) {
        return 0;
    }
    return 4 + type_len + body;
}

/* Split a UR into its parts without copying. Returns false unless the string
   starts with the scheme and has exactly one '/'. */
static bool split(const char *ur, size_t ur_len,
                  const char **type, size_t *type_len,
                  const char **body, size_t *body_len)
{
    if (ur == NULL || ur_len < 4) {
        return false;
    }
    if (lower(ur[0]) != 'u' || lower(ur[1]) != 'r' || ur[2] != ':') {
        return false;
    }

    const char *slash = NULL;
    for (size_t i = 3; i < ur_len; i++) {
        if (ur[i] == '/') {
            if (slash != NULL) {
                return false;   /* a second '/' means multi-part */
            }
            slash = ur + i;
        }
    }
    if (slash == NULL) {
        return false;
    }

    *type = ur + 3;
    *type_len = (size_t)(slash - (ur + 3));
    *body = slash + 1;
    *body_len = (size_t)((ur + ur_len) - (slash + 1));
    return true;
}

bool ur_is_multipart(const char *ur, size_t ur_len)
{
    if (ur == NULL) {
        return false;
    }
    size_t slashes = 0;
    for (size_t i = 0; i < ur_len; i++) {
        if (ur[i] == '/') {
            slashes++;
        }
    }
    return slashes > 1;
}

bool ur_decode(const char *ur, size_t ur_len,
               char *type_out, size_t type_size,
               uint8_t *out, size_t out_size, size_t *out_len)
{
    const char *type, *body;
    size_t type_len, body_len;

    if (type_out == NULL || type_size == 0) {
        return false;
    }
    if (!split(ur, ur_len, &type, &type_len, &body, &body_len)) {
        return false;
    }
    if (!valid_type(type, type_len) || type_len + 1 > type_size) {
        return false;
    }
    if (!ur_bytewords_decode(body, body_len, out, out_size, out_len)) {
        return false;
    }

    for (size_t i = 0; i < type_len; i++) {
        type_out[i] = lower(type[i]);
    }
    type_out[type_len] = '\0';
    return true;
}
